smithy-linters/src/main/java/software/amazon/smithy/linters/MissingSensitiveTraitValidator.java

```diff
@@ -0,0 +1,184 @@
+        }
+    }
+
+    private MissingSensitiveTraitValidator(Config config) {
+        wordMatcher = new WordBoundaryMatcher();
+        if (config.getExcludeDefaults() && config.getTerms().isEmpty()) {
+            //This configuration combination makes the validator a no-op.
+            throw new IllegalArgumentException("Cannot set 'excludeDefaults' to true and leave "
+                    + "'terms' unspecified.");
+        }
+
+        config.getTerms().forEach(wordMatcher::addSearch);
+
+        if (!config.getExcludeDefaults()) {
+            DEFAULT_SENSITIVE_TERMS.forEach(wordMatcher::addSearch);
+        }
+    }
+
+    /**
+     * Finds shapes without the sensitive trait that possibly contain sensitive data,
+     * based on the shape/member name and the list of key words and phrases.
+     *
+     * @param model Model to validate.
+     * @return list of violation events
+     */
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> validationEvents = new ArrayList<>();
+        validationEvents.addAll(scanShapeNames(model));
+        validationEvents.addAll(scanMemberNames(model));
+        return validationEvents;
+    }
+
+    private List<ValidationEvent> scanShapeNames(Model model) {
+        List<ValidationEvent> validationEvents = new ArrayList<>();
+
+        for (Shape shape : model.toSet()) {
+            // Sensitive trait cannot be applied to the 4 types below
+            if (!shape.isMemberShape()
+                    && !shape.isOperationShape()
+                    && !shape.isServiceShape()
+                    && !shape.isResourceShape()
+                    && !shape.hasTrait(SensitiveTrait.class)) {
+                Optional<ValidationEvent> optionalValidationEvent =
+                        detectSensitiveTerms(shape.toShapeId().getName(), shape);
+                optionalValidationEvent.ifPresent(validationEvents::add);
+            }
+        }
+
+        return validationEvents;
+    }
+
+    private List<ValidationEvent> scanMemberNames(Model model) {
+        List<ValidationEvent> validationEvents = new ArrayList<>();
+
+        for (MemberShape memberShape : model.getMemberShapes()) {
+            Shape containingShape = model.expectShape(memberShape.getContainer());
+            Shape targetShape = model.expectShape(memberShape.getTarget());
+
+            if (!containingShape.hasTrait(SensitiveTrait.class) && !targetShape.hasTrait(SensitiveTrait.class)) {
+                Optional<ValidationEvent> optionalValidationEvent =
+                        detectSensitiveTerms(memberShape.getMemberName(), memberShape);
+                optionalValidationEvent.ifPresent(validationEvents::add);
+            }
+        }
+
+        return validationEvents;
+    }
+
+    private Optional<ValidationEvent> detectSensitiveTerms(String name, Shape shape) {
+        Optional<String> matchedTerm = wordMatcher.getAllMatches(name).stream().findAny();
```

Comment:
This looks like another place where I'd just use a for loop and avoid the streaming API since this gets call for tons of times and the streaming API isn't really needed.

```
Set<String> matches = wordMatcher.getAllMatches(name);
return !matches.isEmpty() ? Optional.of(emit(shape, matches)) : Optional.empty();
```

Not sure why `emit` doesn't accept all the matches though in the current implementation. Is it just emitting for the first right now?

Reply:
Since we decided to short-circuit, I'll change this to just getting the first match (as a new method) and emitting that alone

Reply:
Ok, in that case, the method would need a different name than getAllMatches (getFirstMatch or something).

Reply:
Yep, I wrote that method 👍 Current plan is to also leave in getAllMatches as well, unless you think I should remove it since it's not currently used

Reply:
Sure yeah, remove it if it isn't used any longer